// mk-component/lib/components/datagrid/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _omit = require('omit.js');

var _omit2 = _interopRequireDefault(_omit);

var _grid = require('./grid');

var _grid2 = _interopRequireDefault(_grid);

var _fixedDataTable = require('fixed-data-table');

var _cell = require('./cell');

var _cell2 = _interopRequireDefault(_cell);

var _textCell = require('./textCell');

var _textCell2 = _interopRequireDefault(_textCell);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DataGridComponent = function (_React$Component) {
    (0, _inherits3.default)(DataGridComponent, _React$Component);

    function DataGridComponent(props) {
        (0, _classCallCheck3.default)(this, DataGridComponent);

        var _this = (0, _possibleConstructorReturn3.default)(this, (DataGridComponent.__proto__ || (0, _getPrototypeOf2.default)(DataGridComponent)).call(this, props));

        _this.state = {
            height: 0,
            width: 0
        };

        _this.onResize = _this.onResize.bind(_this);
        _this.update = _this.update.bind(_this);
        return _this;
    }

    (0, _createClass3.default)(DataGridComponent, [{
        key: 'componentDidUpdate',
        value: function componentDidUpdate(prevProps, prevState) {
            var dom = _reactDom2.default.findDOMNode(this),
                height = dom.offsetHeight,
                width = dom.offsetWidth;
            if (height != this.state.height || width != this.state.width) {
                this.setState({
                    height: height,
                    width: width
                });
            }
        }
    }, {
        key: 'componentDidMount',
        value: function componentDidMount() {
            this.refreshSize();

            var win = window;
            if (win.addEventListener) {
                win.addEventListener('resize', this.onResize, false);
            } else if (win.attachEvent) {
                win.attachEvent('onresize', this.onResize);
            } else {
                win.onresize = this.onResize;
            }
        }
    }, {
        key: 'componentWillUnmount',
        value: function componentWillUnmount() {
            var win = window;
            if (win.removeEventListener) {
                win.removeEventListener('resize', this.onResize, false);
            } else if (win.detachEvent) {
                win.detachEvent('onresize', this.onResize);
            } else {
                win.onresize = undefined;
            }
        }
    }, {
        key: 'onResize',
        value: function onResize() {
            this.refreshSize();
        }
    }, {
        key: 'refreshSize',
        value: function refreshSize() {
            this.setState({
                height: 0,
                width: 0
            });
        }
    }, {
        key: 'update',
        value: function update() {
            var dom = _reactDom2.default.findDOMNode(this),
                height = dom.clientHeight,
                width = dom.clientWidth;

            this.setState({
                height: height,
                width: width
            });
        }
    }, {
        key: 'render',
        value: function render() {
            var className = (0, _classnames2.default)((0, _defineProperty3.default)({
                'mk-datagrid': true,
                'mk-datagrid-editable': this.props.readonly === false
            }, this.props.className, !!this.props.className));

            return _react2.default.createElement(
                'div',
                { className: className,
                    onKeyDown: this.props.onKeyDown,
                    onKeyUp: this.props.onKeyUp },
                (0, _grid2.default)((0, _extends3.default)({}, (0, _omit2.default)(this.props, ['className']), {
                    width: this.state.width,
                    height: this.state.height
                }))
            );
        }
    }]);
    return DataGridComponent;
}(_react2.default.Component);

DataGridComponent.Cell = _cell2.default;
DataGridComponent.TextCell = _textCell2.default;
DataGridComponent.Column = _fixedDataTable.Column;
DataGridComponent.ColumnGroup = _fixedDataTable.ColumnGroup;

exports.default = DataGridComponent;
module.exports = exports['default'];
