// mk-app-loader/lib/appFactory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends2 = require("babel-runtime/helpers/extends");

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var appFactory = function () {
    function appFactory() {
        (0, _classCallCheck3.default)(this, appFactory);

        this.apps = {};
    }

    (0, _createClass3.default)(appFactory, [{
        key: "registerApp",
        value: function registerApp(name, app) {
            if (this.apps[name]) throw "\u5DF2\u7ECF\u6CE8\u518C\u8FC7\u8FD9\u4E2Aapp\uFF0C\u4E0D\u80FD\u91CD\u590D\u6CE8\u518C. name: " + name;

            this.apps[name] = app;
        }
    }, {
        key: "registerApps",
        value: function registerApps(apps) {
            this.apps = (0, _extends3.default)({}, this.apps, apps);
        }
    }, {
        key: "getApp",
        value: function getApp(name) {
            var app = this.apps[name];

            if (!app) {
                throw "\u6CA1\u6709\u6CE8\u518C\u8FD9\u4E2Aapp. name: " + name;
            }
            return app;
        }
    }, {
        key: "getApps",
        value: function getApps() {
            return this.apps;
        }
    }]);
    return appFactory;
}();

var appFactoryInstance = new appFactory();

exports.default = appFactoryInstance;
module.exports = exports["default"];
